server.js

```diff
@@ -160,63 +160,56 @@ async function setupES(es) {
-      logger.debug(` >> forced upgrade (ENV variable 'ES_UPGRADE' was set) - installing v.${newVersion} ...`);
-      needUpgrade = true;
-    }
-    if (needUpgrade === true) {
-      // Delete any stale index patterns
-      // upgr.delIndexPattern('cicd-perf');
-      // upgr.delIndexPattern('cicd-perf-res');
-      // upgr.delIndexPattern('cicd-perf-errorlog');
-
-      // Import/update latest visualizations and dashboards
-      let importFile = fs.readFileSync('./.kibana_items.json', 'utf8');
-      const replText = nconf.get('env:KB_RENAME');
-      if (replText && typeof replText === 'string') {
-        importFile = importFile.replace(/TIMINGS/g, replText.toUpperCase());
-      }
-      const importJson = JSON.parse(importFile);
-      let esResponse = await es.kbImport(importJson);
-      if (esResponse === true) {
-        await es.defaultIndex(env.INDEX_PERF + '*', '5.6.2')
-      }
-      // upgr.setDefaultIndex(env.INDEX_PERF + '*', '5.6.2');
-
-      // Make sure the latest template is present
-      const templateJson = require('./.es_template.js');
-      templateJson.version = newVersion;
-      await es.putTemplate('cicd-perf', templateJson);
-
-      // Reindexing indices if needed
-      // upgr.reindex('.kibana', '.kibana-perf');
-      // upgr.delIndex('.kibana-perf');
-      // upgr.reindex('cicd-perf-errorlog', 'cicd-errorlog');
-      // upgr.delIndex('cicd-perf-errorlog');
-    } else {
-      logger.debug(` >> no upgrades needed ...`);
-    }
   } catch (err) {
     logger.debug(`Could not reach Elasticsearch Host [${env.ES_HOST}:${env.ES_PORT}] - error message: ${err.message}`);
     nconf.set('env:useES', false);
   }
   return true;
 }
+
+async function checkUpgrade(es, newVer) {
+  let currVer;
+  const currTemplate = await es.getTemplate(env.INDEX_PERF);
+
+  if (currTemplate.hasOwnProperty(env.INDEX_PERF)) {
+    currVer = currTemplate[env.INDEX_PERF].version;
+  }
+
+  if (!currVer || newVer > currVer || nconf.get('es_upgrade') === true ||
+    (currTemplate.hasOwnProperty('error') && currTemplate.error.message === 'Not Found')) {
+
+    doUpgrade(es, currVer, newVer);
+
+  } else {
+    logger.debug(` >> no upgrades needed ... current version: v.${currVer}`);
+  }
+}
+
+async function doUpgrade(es, currVer, newVer) {
+  // Import/update latest visualizations and dashboards
+  logger.debug(` >> Upgrading Elasticsearch to v.${newVer} ... ` +
+    `[Force: ${nconf.get('es_upgrade')} - New: ${!currVer} - Upgr: ${currVer < newVer}]`);
+
+  let importFile = fs.readFileSync('./.kibana_items.json', 'utf8');      // eslint-disable-line no-sync
+  const replText = nconf.get('env:KB_RENAME');
+  if (replText && typeof replText === 'string') {
+    importFile = importFile.replace(/TIMINGS/g, replText.toUpperCase());
+  }
+  const importJson = JSON.parse(importFile);
+  const esResponse = await es.kbImport(importJson);
+  if (esResponse === true) {
+    await es.defaultIndex(env.INDEX_PERF + '*', '5.6.2');
```

Comment:
What does `5.6.2` represent? Magic numbers in code always make me a bit wary.

Reply:
Good catch! It's the version of Elasticsearch. I'm going to move this to the config.